File: problem4/problem.js
// A palindromic number reads the same both ways. 
// The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
// Find the largest palindrome made from the product of two 3-digit numbers.

export const getNumberOfDigitsInNumber = function(number) {
    return Math.max(Math.floor(Math.log10(Math.abs(number))), 0) + 1;
}

export const getLowerRange = function (number) {
    const digits = getNumberOfDigitsInNumber(number)
    return Math.pow(10, digits - 1)
}

export const isPalindrome = function (number) {
    let reverse = 0    
    const original = number
    while (number != 0) {
	    let rem = number % 10
	    reverse = reverse * 10 + rem
        number = Math.floor(number / 10)
    }
    if (reverse === original) {     
        return true
    } else {
        return false
    }
}

export const productCombinations = function (number) {
    const lowerRange = getLowerRange(number)
    let leftNumber = number
    let rightNumber = number
    let maxPalindrome = 0

    while (leftNumber > lowerRange && rightNumber > lowerRange) {
        let candidate = leftNumber * rightNumber
        if (isPalindrome(candidate)) {
            if (candidate > maxPalindrome) {
                maxPalindrome = candidate
            }
        }
        leftNumber--
        if (leftNumber == lowerRange) {
            rightNumber--
            leftNumber = number
        }
    }
    return maxPalindrome
}


// 999 x 999

File: python/problem25/problem.py
import math


def get_number_of_digits_in_number(number):
    return max(math.floor(math.log10(abs(number))), 0) + 1


def solution():
    stop = False
    N_1 = 1
    N_2 = 1
    index = 3
    while stop is False:
        N = N_1 + N_2
        if get_number_of_digits_in_number(N) >= 1000:
            stop = True
        else:
            N_2 = N_1
            N_1 = N
            index += 1

    return index


print(solution())
